pkg/asset/machines/azure/azuremachines.go

```diff
@@ -0,0 +1,247 @@
+			Spec: capi.MachineSpec{
+				ClusterName: clusterID,
+				Bootstrap: capi.Bootstrap{
+					DataSecretName: ptr.To(fmt.Sprintf("%s-%s", clusterID, role)),
+				},
+				InfrastructureRef: v1.ObjectReference{
+					APIVersion: "infrastructure.cluster.x-k8s.io/v1beta2",
+					Kind:       "AzureMachine",
+					Name:       azureMachine.Name,
+				},
+			},
+		}
+
+		result = append(result, &asset.RuntimeFile{
+			File:   asset.File{Filename: fmt.Sprintf("10_machine_%s.yaml", azureMachine.Name)},
+			Object: controlPlaneMachine,
+		})
+	}
+
+	bootstrapAzureMachine := &capz.AzureMachine{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+			Kind:       "AzureMachine",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: capiutils.GenerateBoostrapMachineName(clusterID),
+			Labels: map[string]string{
+				"cluster.x-k8s.io/control-plane": "",
+				"install.openshift.io/bootstrap": "",
+				"cluster.x-k8s.io/cluster-name":  clusterID,
+			},
+		},
+		Spec: capz.AzureMachineSpec{
+			VMSize:                 mpool.InstanceType,
+			Image:                  image,
+			FailureDomain:          ptr.To(mpool.Zones[0]),
+			OSDisk:                 osDisk,
+			AdditionalTags:         tags,
+			AllocatePublicIP:       true,
+			AdditionalCapabilities: additionalCapabilities,
+			SecurityProfile:        securityProfile,
+		},
+	}
+
+	result = append(result, &asset.RuntimeFile{
+		File:   asset.File{Filename: fmt.Sprintf("10_inframachine_%s.yaml", bootstrapAzureMachine.Name)},
+		Object: bootstrapAzureMachine,
+	})
+
+	bootstrapMachine := &capi.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: bootstrapAzureMachine.Name,
+			Labels: map[string]string{
+				"cluster.x-k8s.io/control-plane": "",
+			},
+		},
+		Spec: capi.MachineSpec{
+			ClusterName: clusterID,
+			Bootstrap: capi.Bootstrap{
+				DataSecretName: ptr.To(fmt.Sprintf("%s-%s", clusterID, "bootstrap")),
+			},
+			InfrastructureRef: v1.ObjectReference{
+				APIVersion: "infrastructure.cluster.x-k8s.io/v1beta2",
+				Kind:       "AzureMachine",
+				Name:       bootstrapAzureMachine.Name,
+			},
+		},
+	}
+
+	result = append(result, &asset.RuntimeFile{
+		File:   asset.File{Filename: fmt.Sprintf("10_machine_%s.yaml", bootstrapMachine.Name)},
+		Object: bootstrapMachine,
+	})
+
+	return result, nil
+}
+
+// CapzTagsFromUserTags converts a map of user tags to a map of capz.Tags.
+func CapzTagsFromUserTags(clusterID string, usertags map[string]string) (capz.Tags, error) {
+	tags := capz.Tags{}
+	tags[fmt.Sprintf("kubernetes.io_cluster.%s", clusterID)] = "owned"
+
+	forbiddenTags := sets.NewString()
+	for key := range tags {
+		forbiddenTags.Insert(key)
+	}
+
+	userTagKeys := make([]string, 0, len(usertags))
+	for key := range usertags {
+		userTagKeys = append(userTagKeys, key)
+	}
+	sort.Strings(userTagKeys)
+
+	for _, k := range userTagKeys {
+		if forbiddenTags.Has(k) {
+			return nil, fmt.Errorf("user tags may not clobber %s", k)
+		}
+		tags[k] = usertags[k]
+	}
```

Comment:
Why not use a set here and then report all the conflicting tags at once? If a user has multiple conflicting tags, they'd end up running the installer multiple times before they get rid of all conflicts
```suggestion
	userTagKeys := sets.New[string]()
	for key := range usertags {
		userTagKeys.Insert(key)
	}

    if clobberedTags := userTagKeys.Intersection(forbiddenTags); clobberedTags.Len() > 0 {
		return nil, fmt.Errorf("user tag keys %v are not allowed", sets.List(clobberedTags)) // maybe not needed to sort the keys here; it's only useful for comparison in unit tests
	}
	for _, k := range sets.List(userTagKeys) {
		tags[k] = usertags[k]
	}
```

Reply:
I just picked the code from the installer for this one.
https://github.com/openshift/installer/blob/b8d567115ff6a93ad048808fa4c8a52eb78fe954/pkg/asset/machines/aws/awsmachines.go#L129

Reply:
Aha, I'll send a PR.